# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/layers/PersistentBufferProvider.cpp
# Commit: 7fea32057b92
# Full Hash: 7fea32057b920eac49960541efa2f3fea9074c19
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2022-02-12 09:47:43
# Regressor Bug: 1754130
# File Overlap Count: 2
# Description:
#   Bug 1754130 - Support presenting a WebGLFramebuffer to its own swap chain without opaque FB. r=aosmond,jgilbert
#   
#   Most of the support for presenting a WebGLFramebuffer to a swap chain existed as part of the
#   mechanism for opaque WebXR framebuffer support. However, such "opaque" framebuffer are meant
#   to be opaque in the sense that their attachments can't be inspected or changed, which does
# ==============================================================================

diff -r 1ac8f10e0025 -r 7fea32057b92 gfx/layers/PersistentBufferProvider.cpp
--- a/gfx/layers/PersistentBufferProvider.cpp	Fri Feb 11 15:36:30 2022 +0000
+++ b/gfx/layers/PersistentBufferProvider.cpp	Fri Feb 11 15:36:30 2022 +0000
@@ -10,6 +10,7 @@
 #include "mozilla/layers/TextureClient.h"
 #include "mozilla/layers/TextureForwarder.h"
 #include "mozilla/gfx/gfxVars.h"
+#include "mozilla/gfx/DrawTargetWebgl.h"
 #include "mozilla/gfx/Logging.h"
 #include "mozilla/Maybe.h"
 #include "mozilla/StaticPrefs_layers.h"
@@ -100,15 +101,20 @@
     DrawTarget* aDt)
     : PersistentBufferProviderBasic(aDt) {
   MOZ_COUNT_CTOR(PersistentBufferProviderAccelerated);
+  MOZ_ASSERT(aDt->GetBackendType() == BackendType::WEBGL);
 }
 
 PersistentBufferProviderAccelerated::~PersistentBufferProviderAccelerated() {
   MOZ_COUNT_DTOR(PersistentBufferProviderAccelerated);
 }
 
-ClientWebGLContext* PersistentBufferProviderAccelerated::AsWebgl() {
-  return (ClientWebGLContext*)mDrawTarget->GetNativeSurface(
-      NativeSurfaceType::WEBGL_CONTEXT);
+Maybe<layers::SurfaceDescriptor>
+PersistentBufferProviderAccelerated::GetFrontBuffer() {
+  return static_cast<DrawTargetWebgl*>(mDrawTarget.get())->GetFrontBuffer();
+}
+
+bool PersistentBufferProviderAccelerated::CopySnapshotTo(gfx::DrawTarget* aDT) {
+  return static_cast<DrawTargetWebgl*>(mDrawTarget.get())->CopySnapshotTo(aDT);
 }
 
 static already_AddRefed<TextureClient> CreateTexture(